tablets: support repairing co-located tables using per-table tablet map

mlitvk opened the pull request:
With [Colocated tablets](https://github.com/scylladb/scylladb/pull/22906), the tablet map is shared between multiple colocated tables. Sharing the tablet map is useful because it makes sense for most of the info it holds to be shared - tablet count and token mapping, tablet replicas, resize info, migration info, etc, and it avoids duplication of the data.

However, some info in the tablet map is better being stored per-table, not shared. For example, information about tablet repair - `repair_task_info`, `repair_time`, `repair_scheduler_config`, `sstables_repaired_at`.
Since these fields are shared it means we can't repair a single co-located table independently. In the initial release with co-located tables, due to this limitation, we decided to not allow repair of non-base co-located tables at all (https://github.com/scylladb/scylladb/issues/27119), so the repair information belongs only to the base table in practice.
Now we want to improve this and support repairing non-base co-located tables as well, because a user may want to repair a co-located MV for example, and we want the repair state of the base table and co-located tables to be independent.

To do this, we split the tablet map into two components - shared and per-table.
The shared part continues to hold most of the information and is always written to the base table's partition in `system.tablets`.
The per-table part contains all the repair related fields.
In the tablet metadata, for each table there are two pointers, for the shared `shared_tablet_map` and a `per_table_tablet_map`, and it provides a `tablet_map` interface that hides this by redirecting to the correct component.
The per-table map is written to the partition of each table in `system.tablets`. For colocated tables, previously we had only a static row that points to the base table, and now we have also clustering rows that store the per-table columns. the clustering rows of non-base tables are sparse - we write rows only for tablets that have repair information.

This enables us to support repairing co-located tables.
we allow repair requests for co-located tables and write them to the requested table's per-table map, then the scheduler picks a repair request - it can pick only a single table within a co-location group at a time, and then the topology coordinator runs the repair for the scheduled table.
Note that when the topology coordinator repairs a table, we will move the entire co-location group to `repair` transition stage, because the tablet transition stage is still shared, but it only repairs the requested table.

The per-table map is currently used only for the repair, but it's a general construct that may have more uses in the future for data that we want to store independently for colocated tables instead of being shared by the co-location group.

Fixes https://github.com/scylladb/scylladb/issues/24816

backport not needed - this is an enhancement

PR structure:
1. introduce `per_table_tablet_map`: this will be the per-table part of the tablet map that contains the repair-related fields that we want to be per-table. currently the repair fields are in both the shared and per-table maps, but eventually they will be removed from the shared map.
2. introduce `tablet_map_view`: this contains both the shared part and per-table part. this will replace eventually the current `tablet_map`, and the current `tablet_map` will be renamed to `shared_tablet_map`.
3. start using `tablet_map_view` as the preferred tablet map type in `tablet_metadata`. now each table points to a shared map and a per-table map. when doing operations on table groups, as opposed to a single table, such as tablet migration and resize, we prefer to use the `shared_tablet_map`.
4. read the per-table columns into the per-table maps. (for base tables we read them into both the shared map and the per-table map until it's removed from the shared map.)
5. write the per-table maps as mutations. for non-base co-located tables, we write rows only for tablets that have per-table information.
6. now we want to remove the repair fields from the shared map and have them only in the per-table map. we need first to modify all users of the repair fields to use `tablet_map_view`, then we remove the fields.
7. update the storage_service and topology_coordinator to support repairing a co-located table. allow user requests for co-located tables, and have the topology_coordinator repair the requested table out of the co-location group.
8. rename `tablet_map` -> `shared_tablet_map` and `tablet_map_view` -> `tablet_map`

Example:
1. we have a base table and a co-located table. initially the co-located table has only the static row, it doesn't have any per-table info. it means the 2 tablets of the co-located table take the default values for the per-table fields.
```cql
 table_id                             | base_table                           | last_token          | repair_time
--------------------------------------+--------------------------------------+---------------------+-------------
 b3a945f0-d11c-11f0-99d3-6f2e7e3a4438 |                                 null |                  -1 |        null
 b3a945f0-d11c-11f0-99d3-6f2e7e3a4438 |                                 null | 9223372036854775807 |        null
 b3a96d00-d11c-11f0-99d3-6f2e7e3a4438 | b3a945f0-d11c-11f0-99d3-6f2e7e3a4438 |                null |        null
```
2. repair the tablet of the co-located table with token=-1: now we have a clustering row for the tablet with token=-1 with the per-table values. the second tablet of the co-located table still takes the default values since it doesn't have a row.
```cql
 table_id                             | base_table                           | last_token          | repair_time
--------------------------------------+--------------------------------------+---------------------+---------------------------------
 b3a945f0-d11c-11f0-99d3-6f2e7e3a4438 |                                 null |                  -1 |                            null
 b3a945f0-d11c-11f0-99d3-6f2e7e3a4438 |                                 null | 9223372036854775807 |                            null
 b3a96d00-d11c-11f0-99d3-6f2e7e3a4438 | b3a945f0-d11c-11f0-99d3-6f2e7e3a4438 |                  -1 | 2025-12-04 14:23:41.596000+0000
```

piodul commented:
> @tgrabiec @asias please review.

@asias ping (@tgrabiec is on PTO this week)

mlitvk commented:
Thanks @asias 
I agree we can improve it to be more consistent across the code.
in the PR I changed only the places in the code that access the repair fields to use `get_tablet_map_view`, and all other uses of the tablet map that access only the shared parts remain unchanged. it's mostly to keep the PR smaller and more focused.
I can continue to transform it incrementally, maybe as a follow-up.

asias commented:
The rpc verb TABLET_REPAIR_COLOCATED and associated code like sorage_service::repair_colocated_tablets, and the feature bit if any  should go away.  The code is NOT released in any release yet, so we can and should remove it as soon as possible.

asias commented:
> Thanks @asias I agree we can improve it to be more consistent across the code. in the PR I changed only the places in the code that access the repair fields to use `get_tablet_map_view`, and all other uses of the tablet map that access only the shared parts remain unchanged. it's mostly to keep the PR smaller and more focused. I can continue to transform it incrementally, maybe as a follow-up.

You can work on top of the latest commit in this PR.  The tablet_map and tablet_info interface change are significant, so let's not leave a temporary and half-done state in master.

mlitvk commented:
> The rpc verb TABLET_REPAIR_COLOCATED and associated code like sorage_service::repair_colocated_tablets, and the feature bit if any should go away. The code is NOT released in any release yet, so we can and should remove it as soon as possible.

ok, I removed it. I hope it gets in the next release

> You can work on top of the latest commit in this PR. The tablet_map and tablet_info interface change are significant, so let's not leave a temporary and half-done state in master.

I updated the PR with your suggestions, thanks. I renamed `tablet_map` to `shared_tablet_map` and `tablet_map_view` to `tablet_map`, and similarly with `tablet_info`.

asias commented:
```
    tablets: switch tablet_map alias to tablet_map_view
    
    change `tablet_map` to be an alias of `tablet_map_view` instead of
    `shared_tablet_map`, and fix all compilation issues due to the different
    interfaces.
```

We need to make sure each and every commit compiles for bisecting. You can squash relevant commits into one.

asias commented:
I have concerns about the design of using the tablet transitions of the base table for all the colocated tables. Say we have base table t0 and 100 colocated tables t1 to t100. Now we want to repair all the tables.  Currently, we have to serialize the repair of the colocated tables since they share the transition of the base table.  Alternatively, if the colocated tables use standalone transition, we can repair all the t1 and t100 in parallel. We have similar issues with migration too.

mlitvk commented:
> Say we have base table t1 and two colocated tables t2 and t3. Now we repair t3 only, how do we store the repair_time of t3 in system.tablets? Do we store repair_time just as what we do for a normal and non-colocated table? Please give some concrete examples of the system.tablets tables.

example:
1. we have a base table `t` and two colocated views `vt1`, `vt2`. initially repair_time is null.
the base table has a clustering row for each tablet as normally, and the colocated tables have only a static row with `base_table` set to `t`.
```
 table_name | base_table                           | last_token          | repair_time
------------+--------------------------------------+---------------------+-------------
          t |                                 null |                  -1 |        null
          t |                                 null | 9223372036854775807 |        null
        vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                null |        null
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                null |        null
```
2. `nodetool cluster repair --keyspace ks --table vt1 --tablet-tokens 0`
we repair a single tablet of `vt1` and write `repair_time` to the clustering row in `vt1` for this tablet.
for colocated tables, we have "sparse" clustering rows, only for tablets that have per-table information.
```
 table_name | base_table                           | last_token          | repair_time
------------+--------------------------------------+---------------------+---------------------------------
          t |                                 null |                  -1 |                            null
          t |                                 null | 9223372036854775807 |                            null
        vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                null |                            null
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 | 9223372036854775807 | 2025-08-27 05:52:59.788000+0000
```
3. tablet split:
```
cqlsh> SELECT table_name, base_table, last_token, repair_time FROM system.tablets ;

 table_name | base_table                           | last_token           | repair_time
------------+--------------------------------------+----------------------+---------------------------------
          t |                                 null | -4611686018427387905 |                            null
          t |                                 null |                   -1 |                            null
          t |                                 null |  4611686018427387903 |                            null
          t |                                 null |  9223372036854775807 |                            null
        vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                 null |                            null
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |  4611686018427387903 | 2025-08-27 05:52:59.788000+0000
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |  9223372036854775807 | 2025-08-27 05:52:59.788000+0000
```
4. `nodetool cluster repair --keyspace ks --table t`
```
 table_name | base_table                           | last_token           | repair_time
------------+--------------------------------------+----------------------+---------------------------------
          t |                                 null | -4611686018427387905 | 2025-08-27 06:02:09.513000+0000
          t |                                 null |                   -1 | 2025-08-27 06:02:09.529000+0000
          t |                                 null |  4611686018427387903 | 2025-08-27 06:02:09.514000+0000
          t |                                 null |  9223372036854775807 | 2025-08-27 06:02:09.537000+0000
        vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                 null |                            null
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |  4611686018427387903 | 2025-08-27 05:52:59.788000+0000
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |  9223372036854775807 | 2025-08-27 05:52:59.788000+0000
```
5. `nodetool cluster repair --keyspace ks --table vt2 --tablet-tokens -1`
```
 table_name | base_table                           | last_token           | repair_time
------------+--------------------------------------+----------------------+---------------------------------
          t |                                 null | -4611686018427387905 | 2025-08-27 06:02:09.513000+0000
          t |                                 null |                   -1 | 2025-08-27 06:02:09.529000+0000
          t |                                 null |  4611686018427387903 | 2025-08-27 06:02:09.514000+0000
          t |                                 null |  9223372036854775807 | 2025-08-27 06:02:09.537000+0000
        vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                   -1 | 2025-08-27 06:03:04.164000+0000
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |  4611686018427387903 | 2025-08-27 05:52:59.788000+0000
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |  9223372036854775807 | 2025-08-27 05:52:59.788000+0000
```
6. `nodetool cluster repair --keyspace ks --table vt2 --tablet-tokens -4611686018427387905`
```
 table_name | base_table                           | last_token           | repair_time
------------+--------------------------------------+----------------------+---------------------------------
          t |                                 null | -4611686018427387905 | 2025-08-27 06:02:09.513000+0000
          t |                                 null |                   -1 | 2025-08-27 06:02:09.529000+0000
          t |                                 null |  4611686018427387903 | 2025-08-27 06:02:09.514000+0000
          t |                                 null |  9223372036854775807 | 2025-08-27 06:02:09.537000+0000
        vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 | -4611686018427387905 | 2025-08-27 06:03:45.769000+0000
        vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                   -1 | 2025-08-27 06:03:04.164000+0000
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |  4611686018427387903 | 2025-08-27 05:52:59.788000+0000
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |  9223372036854775807 | 2025-08-27 05:52:59.788000+0000
```

mlitvk commented:
> I have concerns about the design of using the tablet transitions of the base table for all the colocated tables. Say we have base table t0 and 100 colocated tables t1 to t100. Now we want to repair all the tables. Currently, we have to serialize the repair of the colocated tables since they share the transition of the base table. Alternatively, if the colocated tables use standalone transition, we can repair all the t1 and t100 in parallel. We have similar issues with migration too.

the transitions are shared because it's usually for tablet migrations which much be shared for colocated tables, and we can't allow conflicting transitions like migrations with repair.
the only exception is we want to have repair for single tables, maybe in parallel, when there's no tablet migration.

how about if the transition remains shared, but if there are multiple colocated tables with repair requests then the topology coordinator issues the repairs in parallel instead of serializing them?

asias commented:
I understand why vt1 and vt2  look  likes this. But I think this is bad for users who check the system.tablets table for whatever purpose. This breaks the fact each tablet has an entry in the system.tablets table. We can share things internally, e.g. the replicas set per tablet, but it makes more sense to make interface in system.tablets unchanged. That is always present the entry for a tablet no matter it is a base table or a non-base table. 
```
 table_name | base_table                           | last_token          | repair_time
------------+--------------------------------------+---------------------+-------------
          t |                                 null |                  -1 |        null
          t |                                 null | 9223372036854775807 |        null
        vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                null |        null
        vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                null |        null
```

In addition, I think the name base_table does not really reflect the relationship between the base and non-base tables. It is not always like the base and view table in the MV (with real data dependency on the base table). The colocated tables are just having the same token and replicas set. They do not necessarily have data dependency. So the base table is not the base of the non-base tables. 

Sharing tablet_map and tablet_info makes base and non-base table looks very different and inconsistent. Perhaps we can reconsider if the sharing design is really the best option. It is best that the base and no-base tables look exactly the same. We make the tablet schedule aware some tables are colocated, so they can schedule them together. This way, we can make sure the tables are colocated yet make the interface the tablet unchanged, as if the colocated table did not exist. 

What is the worse case that colocated tables are out of sync, e.g., two tablets are supposed on the same host and shard but they are not? Will it only be a performance issue or a correctness issue.

asias commented:
> another idea I consider, maybe we keep the `tablet_transition_stage` shared, but make the `tablet_transition_kind` per-table, and we can use it to indicate which tables we want to repair when in the `repair` stage. when scheduling a repair, we set the shared stage to `repair`, and set transition kind `repair` only to the tables in the group that we want to repair. then in `handle_tablet_migration` we run the repairs in parallel. for other transition kinds like migration that need to be shared we can set it on all the colocated tables, or set it only on the base table and "inherit" it somehow

```
Each tablet has its own state machine for keeping state of transition stored in group0 which is part of the tablet state. It involves
these properties of a tablet:

- replicas: the old replicas of a table (also set when not in transition)
- new_replicas: the new replicas of a tablet which will become current after transition
- stage: determines which replicas should be used by requests on the coordinator side, and which
  action should be taken by the state machine executor.
- transition: the kind of tablet transition (migration, rebuild, etc.). Affects the behavior of stages and actions
  performed in those stages.

enum class tablet_transition_stage {
    allow_write_both_read_old,
    write_both_read_old,
    streaming,
    rebuild_repair,
    write_both_read_new,
    use_new,
    cleanup,
    cleanup_target,
    revert_migration,
    end_migration,
    repair,
    end_repair,
};

enum class tablet_transition_kind {
    // Tablet replica is migrating from one shard to another.
    // The new replica is (tablet_transition_info::next - tablet_info::replicas).
    // The leaving replica is (tablet_info::replicas - tablet_transition_info::next).
    migration,

    // Like migration, but the new pending replica is on the same host as leaving replica.
    intranode_migration,

    // New tablet replica is replacing a dead one.
    // The new replica is (tablet_transition_info::next - tablet_info::replicas).
    // The leaving replica is (tablet_info::replicas - tablet_transition_info::next).
    rebuild,

    // Like rebuild, but instead of streaming data from all tablet replicas,
    // it repairs the tablet and streams data from one replica.
    rebuild_v2,

    // Repair the tablet replicas
    repair,
};
```

The transition_kind  is higher level than the transition_stage. We can not share the lower level but make the higher level unshared. 

The idea of using sharing the tablet_transition_stage and tablet_transition_kind to implement the colocated tables does not sound very promising, at least to me. It breaks the simplicity of tablets, each tablet has its own "everything", transition, replicas, repair time and so on.  It would be nice to keep it if possible. Theoretically, the job of making tablets co-located are the tablet scheduler. It can make tablets colocated or not. Implementing co-location by using the lower level like sharing the transition kind and stage looks like a layer violation. Take the history of this PR as an example, in v1 we hacked at repair_time field level, in v2 we moved up one layer to tablet_info and tablet_map layer to simplify significantly. If we move one layer up to tablet scheduler, we do not even need to modify any of the existing tablet transition layer code.

mlitvk commented:
I changed the repair plan to not schedule multiple repairs of colocated tablets together to simplify the issues around this. so repairs of colocated tablets will always be serialized.
I think it's ok because:
* currently I think it won't schedule multiple colocated repairs anyway because it crosses the streaming load thresholds
* I think this limitation is still better than the previous situation where you must repair all colocated tablets together
* if there will be need we can do changes to support parallel colocated repairs, maybe it will require some schema changes

asias commented:
> > I understand why vt1 and vt2 look likes this. But I think this is bad for users who check the system.tablets table for whatever purpose. This breaks the fact each tablet has an entry in the system.tablets table. We can share things internally, e.g. the replicas set per tablet, but it makes more sense to make interface in system.tablets unchanged. That is always present the entry for a tablet no matter it is a base table or a non-base table.
> 
> I'm not sure which user needs to check system.tablets, but if needed we can create a virtual table that will display full tablet maps for each table. I think the sharing is useful because we can write mutations only once to the base table and not have inconsistencies.

Internally, task managers could use this info to get the tablet related task info. In our tests, we check the system.tables tables in  many places. Users could script around the system tables too. 

> 
> > ```
> >  table_name | base_table                           | last_token          | repair_time
> > ------------+--------------------------------------+---------------------+-------------
> >           t |                                 null |                  -1 |        null
> >           t |                                 null | 9223372036854775807 |        null
> >         vt2 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                null |        null
> >         vt1 | fd0f5470-8309-11f0-b840-fa0066f04ac1 |                null |        null
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In addition, I think the name base_table does not really reflect the relationship between the base and non-base tables. It is not always like the base and view table in the MV (with real data dependency on the base table). The colocated tables are just having the same token and replicas set. They do not necessarily have data dependency. So the base table is not the base of the non-base tables.
> 
> It's "base" in the sense that other tablet maps are dependent on it, it's not necessarily about data dependency, but I can rename it if you have a better suggestion.

The term base and view table is used by MV. Perhaps,  reference table or skeleton table to reflect the fact that the other tables uses it as a reference or a skeleton for tablet replica mappings. 

> 
> > Sharing tablet_map and tablet_info makes base and non-base table looks very different and inconsistent. Perhaps we can reconsider if the sharing design is really the best option. It is best that the base and no-base tables look exactly the same. We make the tablet schedule aware some tables are colocated, so they can schedule them together. This way, we can make sure the tables are colocated yet make the interface the tablet unchanged, as if the colocated table did not exist.
> 
> We discussed it in the original colocated tablets PR and I think we agreed on the sharing approach. actually my first solution was based on the tablet scheduler like you suggest and I changed it, and I think it's simpler with the sharing.

It simplifies the scheduling but complicates the transition layer. In your original PR, with the scheduling solution, does it guarantee base and non-base tables always have  a strict co-location. Was the transition state never shared between non-base and base tables?

> > What is the worse case that colocated tables are out of sync, e.g., two tablets are supposed on the same host and shard but they are not? Will it only be a performance issue or a correctness issue.
> 
> I'm not sure but it may be needed for correctness of LWT

asias commented:
> I changed the repair plan to not schedule multiple repairs of colocated tablets together to simplify the issues around this. so repairs of colocated tablets will always be serialized. I think it's ok because:
> 
> * currently I think it won't schedule multiple colocated repairs anyway because it crosses the streaming load thresholds
> * I think this limitation is still better than the previous situation where you must repair all colocated tablets together
> * if there will be need we can do changes to support parallel colocated repairs, maybe it will require some schema changes

In practice, at least for the base/view table in MV, the management tool is supposed to repair the base table first then, repair the view table. It is fine to not schedule  multiple repairs of colocated tablets together. 

Yes, allowing repairing a individual co-located table is a big improvement. There is no doubt. However, since the change and design of the co-located table will have a long time impact on how the tablet system works. So, it is worth the extra thinking and exploring of different alternative solutions.

mlitvk commented:
* rebase

mlitvk commented:
* rebase

mlitvk commented:
[test_node_failure_during_tablet_migration[write_both_read_new-destination].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/23198/testReport/junit/cluster/test_tablets_migration/debug___test_py___debug___test_node_failure_during_tablet_migration_write_both_read_new_destination__debug_1): looks like https://github.com/scylladb/scylladb/issues/27478

asias commented on gms/feature_service.hh:
```diff
@@ -145,6 +145,7 @@ public:
     gms::feature tablet_repair_scheduler { *this, "TABLET_REPAIR_SCHEDULER"sv };
     gms::feature tablet_merge { *this, "TABLET_MERGE"sv };
     gms::feature tablet_rack_aware_view_pairing { *this, "TABLET_RACK_AWARE_VIEW_PAIRING"sv };
+    gms::feature tablet_map_per_table { *this, "TABLET_MAP_PER_TABLE"sv };
```
The tablet map was per table before co-location tables. Perhaps we'd add co-location to the feature name, e.g.,  PER_TABLE_TABLET_MAP_FOR_COLOCATED_TABLETS.

asias replied:
Using PER_TABLE is very confusing. How about just TABLET_MAP_COLOCATED_TABLETS? This states the feature is about tablet map and colocated tablets.

avikivity commented:
> > I don't see any documentation.
> > How does the user choose which co-located tables to repair?
> 
> the user just requests repair for the table he wants, like before. it doesn't matter if it's a co-located table. there is no documentation because the goal is to make it transparent for the user.

Do we want to repair every collocated table for certain?

asias commented:
> > > > I don't see any documentation.
> > > > How does the user choose which co-located tables to repair?
> > > 
> > > 
> > > the user just requests repair for the table he wants, like before. it doesn't matter if it's a co-located table. there is no documentation because the goal is to make it transparent for the user.
> > 
> > 
> > Do we want to repair every collocated table for certain?
> 
> It really depends on the data dependence between the base and the co-located table. For MVs, we do repair both base and view table.

I think the co-location feature is general that the base and co-located table can choose to either have or not have data dependence at all. Even for a specific base table, the co-located tables could have completely difference data dependence. For example, users might just want to place two tables together and have the same tablet ranges, but the data in the two tables are completely independent.

piodul commented on service/topology_coordinator.cc:
```diff
@@ -2109,14 +2110,16 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                        co_await coroutine::parallel_for_each(replicas, [ms = &_messaging, table_id] (const locator::host_id& h) -> future<> {
+                            co_await ser::repair_rpc_verbs::send_repair_update_repaired_at_for_merge(ms, h, table_id);
+                        });
+                    }
```
Shouldn't the new logic be gated behind the new cluster feature?

piodul commented on replica/tablets.cc:
```diff
@@ -731,6 +717,24 @@ struct tablet_metadata_builder {
             co_await tm.set_colocated_table(table, base_table);
         }
+
+        for (auto& [table, tablet_token] : pending_update_repair_time) {
```
How many elements can `pending_update_repair_time` have? Perhaps we should `maybe_yield` in the loop? Additionally, do we need to clear it gently?

piodul commented on test/boost/tablets_test.cc:
```diff
@@ -817,6 +885,135 @@ SEASTAR_TEST_CASE(test_tablet_metadata_update) {
+
+        // remove a tablet row from the per-table map of table2
+        {
+            testlog.info("MICHAEL remove");
```
Looks like it should be removed?

piodul reviewed (commented):
I think I am more than halfway through with review... Found mostly nits for now, with possibly one missing cluster feature check and some potentially stall-prone loops.

gusev-p commented:
> alternative PR: #28181 it uses different approach to the `tablet_map` unsharing as suggested by @gusev-p, thanks and it currently has only the first part for easier review I can add the other parts later, which should be mostly composed of commits from here

Thanks for taking the time to investigate this alternative.
